packages/flutter_tools/lib/src/build_system/targets/web.dart

```diff
@@ -267,3 +275,103 @@ class WebReleaseBundle extends Target {
 
   }
 }
+
+/// Generate a service worker for a web target.
+class WebServiceWorker extends Target {
+  const WebServiceWorker();
+
+  @override
+  String get name => 'web_service_worker';
+
+  @override
+  List<Target> get dependencies => const <Target>[
+    Dart2JSTarget(),
+    WebReleaseBundle(),
+  ];
+
+  @override
+  List<String> get depfiles => const <String>[
+    'service_worker.d',
+  ];
+
+  @override
+  List<Source> get inputs => const <Source>[];
+
+  @override
+  List<Source> get outputs => const <Source>[];
+
+  @override
+  Future<void> build(Environment environment) async {
+    final List<File> contents = environment.outputDir
+      .listSync(recursive: true)
+      .whereType<File>()
+      .where((File file) => !file.path.endsWith('service_worker.js')
+        && !globals.fs.path.basename(file.path).startsWith('.'))
+      .toList();
+    // TODO(jonahwilliams): determine whether this needs to be made more efficient.
+    final Map<String, String> uriToHash = <String, String>{
+      for (File file in contents)
+        '/${globals.fs.path.relative(file.path, from: environment.outputDir.path)}':
+          md5.convert(await file.readAsBytes()).toString(),
+    };
+    final File serviceWorkerFile = environment.outputDir
+      .childFile('service_worker.js');
+    final Depfile depfile = Depfile(contents, <File>[serviceWorkerFile]);
+    final String serviceWorker = generateServiceWorker(uriToHash);
+    serviceWorkerFile
+      .writeAsStringSync(serviceWorker);
+    depfile.writeToFile(environment.buildDir.childFile('service_worker.d'));
+  }
+}
+
+/// Generate a service worker with an app-specific cache name a map of
+/// resource files.
+///
+/// We embed file hashes directly into the worker so that the byte for byte
+/// invalidation will automatically reactivate workers whenever a new
+/// version is deployed.
+// TODO(jonahwilliams): on re-activate, only evict stale assets.
+String generateServiceWorker(Map<String, String> resources) {
+  return '''
+'use strict';
+const CACHE_NAME = 'flutter-app-cache';
+const RESOURCES = {
+  ${resources.entries.map((MapEntry<String, String> entry) => '"${entry.key}": "${entry.value}"').join(",\n")}
+};
+
+self.addEventListener('install', function (event) {
```

Comment:
According to https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers, activation always follows installation, so if we add all keys to the cache here, then in the activation even we will delete them all and add them again. It seems the only thing that should go in the "install" event is things that do not change from one version of the service worker to the next.

IOW, I think if you delete the "install" event listener entirely the service worker will continue working.

Reply:
install is only triggered when the sw is added, removing 'install' sgtm

Reply:
Done